docs/lang/articles/advanced/performance.md

```diff
@@ -66,3 +66,122 @@ def func():
     for i in range(8192):  # no decorator, use default settings
         ...
 ```
+
+# Local Storage Optimizations
+
+Taichi comes with a few optimizations that leverages the *fast memory* (e.g. CUDA
+shared memory, L1 cache) for performance optimization. The idea is straightforward:
+Where possible, Taichi substitutes the access to the global memroy (slow) with that
+to the local one (fast), and writes the data in the local memory back to the global
+memory in the end. Such transformations preserves the semantics of the original
+program (will be explained later).
+
+## Thread Local Storage (TLS)
+
+TLS is mostly designed to optimize the parallel reduction. When Taichi identifies
+a global reduction pattern in a `@ti.kernel`, it automatically applies the TLS
+optimization to make the generated code match those that are commonly found in
```

Comment:
it automatically applies the TLS optimizations during code generation similar to those found in common GPU reduction implementations.